### PR TITLE
dateBox - show value with invalid hours with internal validation result (T939117)

### PR DESCRIPTION
**Pull request by alexander-kotov-dx**

### CODE REVIEW

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 1]
```diff
@@ -3327,6 +3327,25 @@ QUnit.module('datebox with time component', {
...
+            value: new Date('2015/1/25 14:00:00')
+        });
+        const dateBox = $dateBox.dxDateBox('instance');
+        const $hourDownButton = $(dateBox.content()).find('.dx-numberbox-spin-down').first();
+        const $input = $dateBox.find(`.${TEXTEDITOR_INPUT_CLASS}`);
+
+        $hourDownButton.trigger('dxpointerdown');
+        $(CALENDAR_APPLY_BUTTON_SELECTOR).first().trigger('dxclick');
+
+        assert.strictEqual($input.val(), '13:00', 'input displays a correct value');
```
Comment (DokaRus) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js:3348]: Please add an asserts for the `value` and `isValid` values

File: js/ui/date_box/ui.date_box.base.js [hunk 2]
```diff
@@ -606,8 +610,12 @@ const DateBox = DropDownEditor.inherit({
 
     _applyButtonHandler: function(e) {
         const value = this._strategy.getValue();
-        if(this._applyInternalValidation(value)) {
+        const validationResult = this._applyInternalValidation(value);
+        if(validationResult.isValid) {
             this.dateValue(value, e.event);
+        } else if(validationResult.isDate) {
+            const displayedText = this._getDisplayedText(value);
+            this._input().val(displayedText);
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.base.js]:
```suggestion
            this._renderDisplayText(displayedText);
```

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 3]
```diff
@@ -3327,6 +3327,25 @@ QUnit.module('datebox with time component', {
         assert.equal(dateBox.option('value').getMilliseconds(), 0, 'milliseconds has zero value');
     });
 
+    QUnit.test('invalid (by internal validation) value should be displayed if it is selected by time numberboxes (T939117)', function(assert) {
+        const $dateBox = $('#dateBox').dxDateBox({
+            type: 'datetime',
+            pickerType: 'calendar',
+            opened: true,
+            displayFormat: 'HH:mm',
+            min: new Date('2015/1/25 14:00:00'),
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: Let's specify `date` constant?

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 4]
```diff
@@ -3327,6 +3327,25 @@ QUnit.module('datebox with time component', {
         assert.equal(dateBox.option('value').getMilliseconds(), 0, 'milliseconds has zero value');
     });
 
+    QUnit.test('invalid (by internal validation) value should be displayed if it is selected by time numberboxes (T939117)', function(assert) {
+        const $dateBox = $('#dateBox').dxDateBox({
+            type: 'datetime',
+            pickerType: 'calendar',
+            opened: true,
+            displayFormat: 'HH:mm',
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js:3338]: Is this prop necessary?

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 5]
```diff
@@ -3327,6 +3327,25 @@ QUnit.module('datebox with time component', {
...
+            min: new Date('2015/1/25 14:00:00'),
+            value: new Date('2015/1/25 14:00:00')
+        });
+        const dateBox = $dateBox.dxDateBox('instance');
+        const $hourDownButton = $(dateBox.content()).find('.dx-numberbox-spin-down').first();
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: Let's specify global constant for `'.dx-numberbox-spin-down'` and refactor similar code places

File: js/ui/date_box/ui.date_box.base.js [hunk 6]
```diff
@@ -530,7 +530,11 @@ const DateBox = DropDownEditor.inherit({
             }
         });
 
-        return isValid;
+        return {
+            isValid,
+            isDate,
+            isDateInRange
```
Comment (ksercs) [inline js/ui/date_box/ui.date_box.base.js]: Is it necessary to return it?

File: js/ui/date_box/ui.date_box.base.js (hunk 2 above)
Comment (ksercs) [inline js/ui/date_box/ui.date_box.base.js]: Is it ok that `text` option doesn't update here?

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js (hunk 4 above)
Comment (alexander-kotov-dx) [reply testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js:3338]: it is not necessary, but it gives short understandable texts

File: js/ui/date_box/ui.date_box.base.js [hunk 7]
```diff
@@ -606,8 +609,13 @@ const DateBox = DropDownEditor.inherit({
 
     _applyButtonHandler: function(e) {
         const value = this._strategy.getValue();
-        if(this._applyInternalValidation(value)) {
+        const validationResult = this._applyInternalValidation(value);
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.base.js]:
```suggestion
        const { isValid, isDate } = this._applyInternalValidation(value);
```

**Review by ksercs — DISMISSED** [review]

**Review by ksercs — APPROVED** [review]

**Review by San4es — APPROVED** [review]

**Pull request merged**